Documentation/guides/internals/debug-jni-objrefs.md

```diff
@@ -0,0 +1,400 @@
+
+In order to diagnose and fix JNI Object Reference usage bugs, JNI Object Reference logs must first be collected.
+There are two sets of logs: JNI *Local* Reference logs, and JNI *Global* Reference logs.  JNI Local Reference logs
+are *incredibly* voluminous; they should be collected only as a last resort.
+
+There are two ways to obtain JNI Object Reference logs:
+
+ 1. [Collect the *complete* JNI Object Reference log output](#collect-complete-logs), or
+ 2. [Collect "best effort" JNI Object Reference log output](#collect-best-effort-logs).
+
+
+<a name="collect-complete-logs"></a>
+
+### Collect Complete JNI Object Reference Logs
+
+To collect *complete* JNI object reference logs, your app must be "debuggable": the
+[`//application/@android:debuggable`](https://developer.android.com/guide/topics/manifest/application-element#debug) attribute
+within `AndroidManifest.xml` must have the value `true`.  This is typically the case for Debug configuration builds, and
+*not* for Release configuration builds.  (The Google Play Store requires that submitted apps *not* be debuggable.)
+
+Enable JNI Global Reference log collection by setting the `debug.mono.log` system property to a value which contains `gref`:
+
+	adb shell setprop debug.mono.log gref
+
+Then run your app again and trigger the crash.  Once tha app has exited, run:
+
+	adb shell run-as @PACKAGE-NAME@ cat files/.__override__/grefs.txt > grefs.txt
+
+where `@PACKAGE-NAME@` is the value of the [`/manifest/@package` attribute](https://developer.android.com/guide/topics/manifest/manifest-element#package)
+within `AndroidManifest.xml`.  This is generally the filename before `-Signed.apk` in your `bin` directory.
+
+To collect JNI Local Reference logs, the `debug.mono.log` system property should contain `lref`, and the required file is `lrefs.txt`.
+
+Both Local and Global JNI Object References can be collected at the same time:
+
+	adb shell setprop debug.mono.log lref,gref
+
+	# run the app, then
+
+	adb shell run-as @PACKAGE-NAME@ cat files/.__override__/grefs.txt > grefs.txt
+	adb shell run-as @PACKAGE-NAME@ cat files/.__override__/lrefs.txt > lrefs.txt
+
+<a name="collect-best-effort-logs"></a>
+
+### Collect Best Effort JNI Object Reference Logs
+
+Complete JNI Object Reference log collection can only be done for debuggable apps.  If your app isn't debuggable,
+or the crash doesn't reproduce in a debuggable app, then you will need to try for "Best Effort" collection by
+setting the `debug.mono.log` system property to contain either/both `gref+` for Global References and
+`lref+` for Local References:
+
+	adb shell setprop debug.mono.log lref+,gref+
+
+Then, begin collecting `adb logcat` *before* launching your app:
+
+	adb logcat > log.txt
+
+Once your app crashes, `log.txt` will contain the JNI Object Reference logs.
+
+*However*, the output *may* be incomplete.  It is not unusual for information to be missing, because *so much*
+data is written to `adb logcat`.
```

Comment:
We should find out what's the maximum admissible logcat buffer size and suggest setting it here